### PR TITLE
Modified Layers that gives better Non-Seq model support, automatic i/p-o/p dimensions calculations

### PR DESCRIPTION
**Pull request by pranv**
Adds better support for non- sequential models.
Each layer now has 3 default attributes: name(name of the layer), prev(name of the previous layer) and input_dim(size of the input)
Also added Pooling with strides, Zero Padding and Local Response Normalization layer.

1. Merge is a proper layer now[instead of a model]
2. Fork should work and should be fast based on theano optimisations
3. Automatic output dimensions are calculated and stored for each layer based on the input dim. Each succeeding layer can then use that info to set its input size.

Examples:
```
model = Sequential()

model.add(Convolution2D(32, 3, 3, border_mode='full', input_dim=(3,32,32)))
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(2, 2)))
model.add(Dropout(0.25))

model.add(Convolution2D(64, 3, 3, border_mode='full'))
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense( 512, init='normal'))
model.add(Activation('relu'))
model.add(Dropout(0.5))

model.add(Dense( 10, init='normal'))
model.add(Activation('softmax'))
```

Non Seqential:

```
model = Sequential()
model.add(Dense(20,5, name='input'))
model.add(Dense(5, name='pro1', prev='input'))
model.add(Dense(5, name='pro2', prev='input'))
model.add(Merge(prev=['pro1', 'pro2']))

```

Didnt have enough time to test all the code properly. Hopefully we all can test and fix the issues.
Caffe integration is on the way[just need a couple of days]

The changes have not been applied to Recurrent and Embeddings [not a pro in them, so left it], hopefully someone can do it

### CODE REVIEW

**Comment by pranv** [comment]
The commit history is not accurate. I was working based on the old code[2 months ago], quite a lot has changed since then, so I just copied and pasted my code in the newest version. I apologize for that.

File: keras/layers/core.py [hunk 1]
```diff
@@ -16,21 +16,65 @@
...
+    def setup(self):
+        '''
+            Contains code that depends on values that are optional but can be understood from context (ex: input_dim)
+        '''
+        return
+
+    def connect(self, node):
+        self.previous = node
+
+        if self.input_dim == (None, ):
```
Comment (iskandr) [inline keras/layers/core.py]: What's the difference between `self.input_dim` being `(None,)` and `None`? Could it also be some combination of `None` and other values, such as `(128, None)`?

File: keras/layers/core.py [hunk 2]
```diff
@@ -16,21 +16,65 @@
...
+        self.output_dim = (None, )
+
+    def setup(self):
+        '''
+            Contains code that depends on values that are optional but can be understood from context (ex: input_dim)
+        '''
+        return
+
+    def connect(self, node):
+        self.previous = node
```
Comment (iskandr) [inline keras/layers/core.py]: It's a little confusing to have both `self.prev` and `self.previous`, what's the difference between them?

Comment (pranv) [inline keras/layers/core.py]: As a standard, all dimensions are passed around as tuples. It can be in any combination, but the calculations should hold good through the layers.



self.prev is the name of the previous layer(as a string)


self.previous is the actual previous layer object.

Comment (iskandr) [inline keras/layers/core.py]: >self.prev is the name of the previous layer(as a string)
>self.previous is the actual previous layer object 

Maybe worth commenting boldly and/or naming them more distinctly? Or, perhaps, could the two just be unified? Why can't a user pass a layer object directly in place of a name?

>As a standard, all dimensions are passed around as tuples. It can be in any combination, but the calculations should hold good through the layers.

I'm not sure if that's answering the question about `input_dim`; if so, then I'm still confused about what it means for `input_dim` to be `None` and why `connect` only checks for `(None,)` as opposed to e.g. `(None, None)`.

**Comment by fchollet** [comment]
Any way you could break this down into key features submitted into separate PRs, such as:
- merge / fork (merge was already a layer, btw).
- dimensionality inference
- Caffe models compatibility

Such a massive PR will be quite difficult to review and eventually merge. Unbundling it will be the only practical way to move forward with these features.

File: keras/caffe/converter.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			# TANH
+			model.add(Activation('tanh'))
+		else:
+			print "The Layer ", layer_type_nb[layer_config.type], " is not currently supported or is out of context"
+
+def ConvertWeights():
+
+def ConvertSolver():
+
+def ConvertMean():
```
Comment (mthrok) [inline keras/caffe/converter.py]: Aren't `pass` missing here? Or does this work without `pass`?

**Comment by mthrok** [comment]
And about caffe model compatibility, looking at their filter visualization example, [here] (http://nbviewer.ipython.org/github/BVLC/caffe/blob/master/examples/filter_visualization.ipynb
) I think all what we need to do for interacting with caffe mode is just to use their `pycaffe` module with `import caffe`. So adding `caffe_pb2` file generated by protocol buffer compiler at caffe building time is not a good idea.

**Comment by fchollet** [comment]
> all what we need to do for interacting with caffe mode is just to use their pycaffe module with import caffe. So adding caffe_pb2 file generated by protocol buffer compiler at caffe building time is not a good idea.

@pranv have you investigated pycaffe? What are your thoughts on this?

**Comment by pranv** [comment]
I'm working on getting Models in Caffe to be converted to the equivalent keras ones in 1 step. Caffe installation is not really a pleasure(its the most complicated DL to install actually - from a end user perspective), so instead by using Google Protocol Buffers and caffe_pb2, I have worked on it. Not sure why @mthork thought it was not a good idea. The couple of people, who have used Caffe through pycaffe I spoke with regarding this also had the same opinion.


Just pass the .caffemodel file and the .prototxt model description file, you'll get the same model in keras - ready to compile and give you predictions. You can instantly use all the models trained on caffe so far. One can fine tune them, build on them etc.,. It's not about just filter visualization.




I've also processed the caffe models as a proper Directed Acyclic Graph. Thus, it applies to all sorts of networks - Siamese, multiple output etc.,




Lastly, the assign weights method of the model, makes all this work for embedding a pretrained caffe network inside our own network. The best use case I can think of for this is Karpathy et al. image Description networks.




Regarding splitting this Pull Request:

I agree that it has to be split up. All the things in this PR came off while working on Caffe Integration as a side effect and I kept working on them, finally ending up here. And in some way, they are all interrelated.




I'll close this request, and submit new ones.




I was hoping to get some feedback about the features themselves. So, if there are any issues I can fix them and submit them, rather than work twice.

**Comment by mthrok** [comment]
I am sure that everyone using keras love the idea of converting caffe model into keras model with one function, and I think your `caffe/converter.py` script is good starting point. 

What I wanted to say is that 'adding `caffe_pb2.py` to keras code base is not a good idea'. If it sounded confusing, I'm sorry. 
The point is that `caffe_pb2.py` is dynamically generated by `protocol buffer compiler` when `caffe` is built. (Check out the first build message when you build `caffe` with `cmake`) So we are not sure if it is always same file for everyone with different system configuration. (different OS, protobuf, caffe, C++ compiler version, etc) And it is possible that your `caffe_pb2.py` does not work in my environment or other that of people's.

I understand installing caffe is sometimes not easy. When I got first my Yosemite, brew didn't have `protobuf` package so I had to do it manually. But I think now `brew` has it and other package managers too (I tried with `apt-get` with Ubuntu 14.04, and it instantly installed everything.) But either way it will not be a valid reason to add compiler-generated code to `keras` code base. 

Plus, as long as you install `caffe` and set a collect path to `pycaffe` (in default it is `<caffe_root>/build/install/python` `$PYTHONPATH`, then you do have access to the exact `caffe_pb2.py` file. With this approach we can continue developing our `keras-caffe` module without being worried about `caffe`'s development status.

**Comment by fchollet** [comment]
> I'm working on getting Models in Caffe to be converted to the equivalent keras ones in 1 step. 

That sounds good!

> What I wanted to say is that 'adding caffe_pb2.py to keras code base is not a good idea'.

 It might be burdensome to add this to the codebase indeed. It's not clear that it belongs in the Keras codebase.

So, what would be a way to:
- reduce what gets added to the codebase to a minimum
- reduce external dependencies to a minimum?

(I have no strong opinion on this because I have not investigated the problem).

> I was hoping to get some feedback about the features themselves. So, if there are any issues I can fix them and submit them, rather than work twice.

I will definitely review this in the next couple days and give you my feedback. I think these are very important features and we'll do everything we can to make sure they fit nicely into the Keras project and codebase.

**Comment by pranv** [comment]
You're right when you say that Protocol buffers are compiled. But this not an OS/env specific compilation. According to Google Protobuf Documentation:

> Once you've defined your messages, you run the protocol buffer compiler for your application's language on your .proto file to generate data access classes. These provide simple accessors for each field, as well as methods to serialize/parse the whole structure to/from raw bytes – so, for instance, if your chosen language is C++, running the compiler on the above example will generate a class called Person. You can then use this class in your application to populate, serialize, and retrieve Person protocol buffer messages.

Since we are using platform agnostic language, I believe that it won't cause any issues.

If adding it to the code base is an issue(which I believe is not), I can write a bash script to fetch it from the web on demand the first time.

**Comment by mthrok** [comment]
If you are going to host protocol file, then you have to update protocol every time official caffe modify it. Here is how frequently they modify the protocol definition.
https://github.com/BVLC/caffe/commits/master/src/caffe/proto/caffe.proto

If you are going to set up bash script, I think fetching the latest `caffe.proto` from the official and run protobuf compiler in local environment requires less maintenance cost.

**Comment by pranv** [comment]
Well, if the protocol is updated in such a way that we can no longer use it, it also means that all old Caffe Models can't be used, which won't help our purpose of bringing the rich and already existing pre-training nets ecosystem from caffe to keras and theano world. In this case, having caffe_pb2 in our code actually helps.

If the protocol doesn't break any things, we are well and good. Maybe, like you stated, we need to change it to align with the latest caffe version. This also means that we need to change the accompanying keras code, hence doing the same together helps and is a must.

This is my current thinking at least. I primarily like this approach because getting caffe in its entirety to work is a real big pain. However, if you are experienced in caffe and keras, and you think that using pycaffe is better, I'll take your advice and work on that :)

File: keras/layers/core.py (hunk 1 above)
Comment (fchollet) [inline keras/layers/core.py]: `(None,)` seems rather confusing. Why not `None`?

File: keras/layers/core.py [hunk 4]
```diff
@@ -75,79 +119,40 @@ def get_params(self):
...
+            output_dim[-1] = sum(sizes)
+            return tuple(output_dim)
 
 class Dropout(Layer):
     '''
         Hinton's dropout.
     '''
-    def __init__(self, p):
-        super(Dropout,self).__init__()
+    def __init__(self, p, name=None, prev=None, input_dim=(None, )):
```
Comment (fchollet) [inline keras/layers/core.py]: I think this setup brings a lot of code redundancy. Why not call `super` with the keyword arguments to `__init__`?

**Comment by fchollet** [comment]
Why are input sizes passed in constructor arguments? Why not something like `.get_input_shape()` (similar to `.get_input()`, etc).

**Comment by fchollet** [comment]
Can you describe the key architecture changes you're making, at a high level, and what your motivations are for these changes?
